CDCgov/prime-reportstream
# 2598 Implemented the Lookup Tables API

This PR adds the lookuptable API and changes the CLI to use the API. Use `./prime lookuptables --help` and `./prime lookuptables <command> --help` for all options. The commands are:

**Test Steps:**
These steps use the fips-county table data since it is already available and we can relate it to how other lookup tables work, however, the fips-county table will stay as a file based table for the time being.  This tool will be used first for tables needed by the Sender Automation team.
1. Build the branch by running `gradle package` and run RS.
2. Run `./prime lookuptables --help` and verify the help text with the available commands is displayed.
3. Run `./prime lookuptables list` to get the list of tables and verify no tables as found as the database is currently empty.
4. Load a new table by running the command `./prime lookuptables create -n fips-county -i ./metadata/tables/fips-county.csv`.  Verify that the data is shown in table format.  Answer the prompt with 'Y' and verify the table was created.
5. Run `./prime lookuptables list -a` and verify the fips-county table version 1 is present and active.
6. Copy the ./metadata/tables/fips-county.csv file to a temporary location and delete a few lines and change others.  This will make a revised table for our test.
7. Create a new version of the table by running `./prime lookuptables create -n fips-county -i <your temp file>`.  Verify that the data is shown in table format and a diff is shown.  Answer the prompt with 'Y' and verify the table was created.
8. Run `./prime lookuptables list -a` and verify that both table versions are shown with version 1 still active.
9. Activate the new table by running `./prime lookuptables  activate -n fips-county -v 2` and answer 'Y' on the prompt.
10. Run `./prime lookuptables list -a` and verify that both table versions are shown and version 2 is now active.
11. Compare table versions by running `./prime lookuptables diff -n fips-county -v1 1 -v2 2` and verify the changes you made are shown as a diff.
12. Run `./prime lookuptables get -n fips-county -v 2` and verify version 2 of the table is displayed as a table on the screen.
15. Run `./prime lookuptables get -n fips-county -v 2 -o ./fips-output-test.csv` and verify a file is written fips-output-test.csv in CSV format with the table data for version 2 of the table.

## Changes
- Created new Lookup Tables API
- Updated CLI to use API
- Created new Environment class, but will integrate that in existing code in another issue to keep PR small.
- Unit tests
- Other minor code cleanups

## Checklist

### Testing

### Process

## Fixes
- #2598

## To Be Done
- #2714 - Cleanup of the other CLI command classes to use the new Environment class and not have classes reference SettingsCommands for non-setting code.

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
